compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/BIRGen.java

```diff
@@ -2232,17 +2415,36 @@ public void visit(BLangTableConstructorExpr tableConstructorExpr) {
     public void visit(BLangSimpleVarRef.BLangTypeLoad typeLoad) {
         BType type = typeLoad.symbol.tag == SymTag.TYPE_DEF ?
                 ((BTypeDefinitionSymbol) typeLoad.symbol).referenceType : typeLoad.symbol.type;
-        visitTypedesc(typeLoad.pos, type, Collections.emptyList());
+        createNewTypedescInst(type, typeLoad.pos);
     }
 
-    private void visitTypedesc(Location pos, BType type, List<BIROperand> varDcls) {
-        BIRVariableDcl tempVarDcl =
-                new BIRVariableDcl(symTable.typeDesc, this.env.nextLocalVarId(names), VarScope.FUNCTION, VarKind
-                        .TEMP);
-        this.env.enclFunc.localVars.add(tempVarDcl);
-        BIROperand toVarRef = new BIROperand(tempVarDcl);
-        setScopeAndEmit(new BIRNonTerminator.NewTypeDesc(pos, toVarRef, type, varDcls));
-        this.env.targetOperand = toVarRef;
+    private void createNewTypedescInst(BType type, Location position) {
+        List<BIROperand> varDcls = mapToVarDcls(type);
+        BTypeSymbol typeSymbol = type.tsymbol;
+        BIRVariableDcl tempVarDcl;
+        BIROperand toVarRef;
+        BIRNonTerminator.NewTypeDesc newTypeDesc;
+        if (this.env.enclBB == null) {
+            if (type.tag == TypeTags.RECORD || typeSymbol.name != Names.EMPTY) {
+                return;
+            }
+            tempVarDcl = new BIRGlobalVariableDcl(symTable.typeDesc, this.env.nextTypedescId(names), VarScope.GLOBAL,
```

Comment:
Let's define this a local variable inside if block, then the casting is not needed.